lib/Runtime/Base/FunctionBody.cpp

```diff
@@ -7898,37 +7903,131 @@ namespace Js
+        Assert(this->GetState() == CodeGenRecorded || this->GetState() == CodeGenDone || this->isAsmJsFunction);
+
+        // !! this is illegal, however (by design) `IsInNativeAddressRange` (right above) needs it
+        return reinterpret_cast<DWORD_PTR>(this->GetNativeEntryPointData()->GetNativeAddress());
+    }
+
+    Js::JavascriptMethod EntryPointInfo::GetThunkAddress() const
+    {
+        Assert(this->GetState() == CodeGenRecorded || this->GetState() == CodeGenDone);
+
+        return this->GetNativeEntryPointData()->GetThunkAddress();
+    }
+
+    Js::JavascriptMethod EntryPointInfo::GetNativeEntrypoint() const
     {
-        Assert(this->state == CodeGenPending);
+        Assert(this->GetState() == CodeGenRecorded || this->GetState() == CodeGenDone || this->isAsmJsFunction);
+        Js::JavascriptMethod thunkAddress = this->GetNativeEntryPointData()->GetThunkAddress();
+        return thunkAddress ? thunkAddress : this->GetNativeEntryPointData()->GetNativeAddress();
+    }
+
+    ptrdiff_t EntryPointInfo::GetCodeSize() const
+    {
+        Assert(this->GetState() == CodeGenRecorded || this->GetState() == CodeGenDone);
+        return this->GetNativeEntryPointData()->GetCodeSize();
+    }
+
+    JitTransferData * EntryPointInfo::GetJitTransferData()
+    {
+        return this->GetNativeEntryPointData()->GetJitTransferData();
+    }
+
+    SmallSpanSequence* EntryPointInfo::GetNativeThrowSpanSequence() const
+    {
+        Assert(this->GetState() != NotScheduled);
+        Assert(this->GetState() != CleanedUp);
+        return this->GetNativeEntryPointData()->GetNativeThrowSpanSequence();
+    }
+
+    void EntryPointInfo::SetNativeThrowSpanSequence(SmallSpanSequence* seq)
+    {
+        Assert(this->GetState() == CodeGenQueued);
+        Assert(this->GetNativeEntryPointData()->GetNativeThrowSpanSequence() == nullptr);
+
+        this->GetNativeEntryPointData()->SetNativeThrowSpanSequence(seq);
+    }
 
-        this->weakFuncRefSet = this->EnsureWeakFuncRefSet(recycler);
-        this->weakFuncRefSet->AddNew(weakFuncRef);
+    uint EntryPointInfo::GetFrameHeight()
+    {
+        return this->GetNativeEntryPointData()->GetFrameHeight();
     }
 
-    EntryPointInfo::WeakFuncRefSet *
-    EntryPointInfo::EnsureWeakFuncRefSet(Recycler *recycler)
+    bool EntryPointInfo::HasInlinees()
     {
-        if (this->weakFuncRefSet == nullptr)
+        return this->GetFrameHeight() > 0;
+    }
+
+    NativeEntryPointData * EntryPointInfo::EnsureNativeEntryPointData()
+    {
+        NativeEntryPointData * data = this->nativeEntryPointData;
+        if (data == nullptr)
         {
-            this->weakFuncRefSet = RecyclerNew(recycler, WeakFuncRefSet, recycler);
+#if ENABLE_OOP_NATIVE_CODEGEN
+            if (JITManager::GetJITManager()->IsOOPJITEnabled())
+            {
+                data = RecyclerNew(this->GetScriptContext()->GetRecycler(), OOPNativeEntryPointData);
```

Comment:
Should these allocations be zeroing?

Reply:
I explicitly initialize all fields in the constructor.